# Map, Filter, Reduce, Lambda & Recursion

## Tasks Today:

1) <b>Lambda Functions</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Saving to a Variable <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Multiple Inputs <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Passing a Lambda into a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; e) Returning a Lambda from a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; f) In-Class Exercise #1 <br>
2) <b>Map</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Map <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #2 <br>
3) <b>Filter</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Filter <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #3 <br>
4) <b>Reduce</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Reduce <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #4 <br>
5) <b>Recursion</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Implementing a Base <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Writing a Factorial Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #5 <br>
6) <b>Generators & Iterators</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Yield Keyword <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Inifinite Generator <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #6 <br>
7) <b>Exercises</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Exercise #1 - Filtering Empty Strings <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Exercise #2 - Sorting with Last Name <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Exercise #3 - Conversion to Farhenheit <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Exercise #4 - Fibonacci Sequence <br>

## Lambda Functions <br>
<p>Lambda functions... or "Anonymous Functions" are referring to inline functions with no name. The keyword lambda denotes the no name function, and executes within a single line. Without saving it to a variable; however, it is not able to be used, unless passed in either as a paramater or within list comprehension.<br>Written as "(keyword lambda) (one or more inputs) (colon) (function to be executed)"</p>

#### Syntax

In [1]:
def add_two(num):
    return num +2

add_two(4)

6

In [4]:
(lambda num: num + 2)(6)

8

In [5]:
(lambda num: num ** 2)(3)

9

In [6]:
(lambda num: 3 - num)(2)

1

In [8]:
result = (lambda num: num - 3)(2)
print(result)

-1


In [9]:
a_list = [1,2,3,4,5]

[(lambda x: x**2)(num) for num in a_list]

[1, 4, 9, 16, 25]

#### Saving to a Variable

In [11]:
square_nums = lambda x: x ** 2

In [12]:
square_nums(2)

4

In [13]:
square_nums(4)

16

#### Multiple Inputs

In [14]:
(lambda num1, num2, num3: num1 + num2 + num3)(4,5,6)

15

#### Passing a Lambda into a Function

In [15]:
def double_numbers(a_list):
    for i in range(len(a_list)):
        a_list[i] = a_list[i] *2
        
    return a_list

double_numbers([1,2,3,4,5])

[2, 4, 6, 8, 10]

In [18]:
def half_numbers(a_list):
    for i in range(len(a_list)):
        a_list[i] = a_list[i] / 2
        
    return a_list

half_numbers([1,2,3,4,5])

[0.5, 1.0, 1.5, 2.0, 2.5]

In [19]:
def increment_numbers(a_list):
    for i in range(len(a_list)):
        a_list[i] = a_list[i] + 1
        
    return a_list

increment_numbers([1,2,3,4,5])

[2, 3, 4, 5, 6]

In [20]:
def modify_numbers(a_list, f):
    for i in range(len(a_list)):
        a_list[i] = f(a_list[i])
        
    return a_list

In [23]:
#double numbers
modify_numbers([1,2,3,4,5], lambda x: x * 2)
#half numbers
modify_numbers([1,2,3,4,5], lambda x: x / 2)
#increment numbers
modify_numbers([1,2,3,4,5], lambda x: x + 1)

[2, 3, 4, 5, 6]

In [25]:
#square numbers, subtract 1, then add 4
modify_numbers([1,2,3,4,5], lambda x: (x ** 2) - 1 + 4)

[4, 7, 12, 19, 28]

#### Returning a Lambda from a Function

In [ ]:
lambda num1, num2: (num1 + num2 + 4) ** 2

In [ ]:
lambda num1, num2: (num1 + num2 + 7) ** 3

In [27]:
def build_lambda(adder, power):
    return lambda num1, num2: (num1 + num2 + adder) ** power

In [29]:
build_lambda(7, 3)(4,5)

4096

#### If Statements within Lambdas

In [30]:
#square every other number in a list
a_list = [1,2,3,4,5]

modify_numbers(a_list, lambda x: x ** 2 if x%2 == 1 else x)

[1, 2, 9, 4, 25]

#### In-Class Exercise #1 <br>
<p>Write an anonymous function that cubes the arguments passed in and assign the anonymous function to a variable 'f'.</p>

In [35]:
f = (lambda x: x**3)

print(f(13))
print(f(144))
print(f(27))

2197
2985984
19683


## Map <br>
<p>The map function allows you to iterate over an entire list while running a function on each item of the list. This is why the map function works well with lambda's, because it simplifies things and you write less lines of code.<br>The syntax for a map function is "map(function to be used, list to be used)"<br>However, you must be careful, as the map function returns a map object, not a list. To turn it into a list we use the list() type conversion.</p>

#### Syntax

In [36]:
a_list = [1,2,3,4,5]

print(list(map(lambda x: x * 2, a_list)))
print(list(map(lambda x: x * 3, a_list)))
print(list(map(lambda x: x * 4, a_list)))
print(list(map(lambda x: x ** 2, a_list)))

[2, 4, 6, 8, 10]
[3, 6, 9, 12, 15]
[4, 8, 12, 16, 20]
[1, 4, 9, 16, 25]


#### Using Lambda's with Map

#### In-Class Exercise #2 <br>
<p>Use the map function to double each number and minus it by one in the list by using a lambda function</p>

In [38]:
a_list = [1,2,3,4,5]

print(list(map(lambda x: x * 2 - 1, a_list)))

[1, 3, 5, 7, 9]


## Filter() <br>
<p>Filter's are similar to the map function, where you're able to pass a function argument and a list argument and filter out something from the list based on the conditions passed. Similar to the map function, it returns a filter object, so you need to type convert it to a list()</p>

#### Syntax

In [39]:
a_list = [1,2,3,4,5,6,7,8,9,10]

list(filter(lambda num: num % 2 == 0, a_list))

[2, 4, 6, 8, 10]

#### Using Lambda's with Filter()

In [40]:
def is_odd(num):
    if num % 2 == 0:
        return True
    else:
        return False
    
list(filter(is_odd, a_list))

[2, 4, 6, 8, 10]

In [ ]:
my_list = ['a,b,c']

#### In-Class Exercise #3 <br>
<p>Filter out all the numbers that are below the mean of the list.<br><b>Hint: Import the 'statistics' module</b></p>

In [41]:
import statistics as stats

a_list = [1,2,3,4,5,6,7,8,9,10]

list(filter(lambda num: num >= stats.mean(a_list), a_list ))

[6, 7, 8, 9, 10]

## Reduce() <br>
<p>Be very careful when using this function, as of Python 3 it's been moved to the 'functools' library and no longer is a built-in function.<br>The creator of Python himself, says to just use a for loop instead.</p>

#### Syntax

In [43]:
from functools import reduce

a_list = [1,2,3,4,5]

def add_nums(num1, num2):
    print(num1, num2)
    return num1 + num2

reduce(add_nums, a_list)

1 2
3 3
6 4
10 5


15

In [44]:
def multiply_nums(num1, num2):
    return num1 * num2

reduce(multiply_nums, a_list)

120

#### Using Lambda's with Reduce()

In [45]:
reduce(lambda num1, num2: num1 * num2, a_list)

120

#### In-Class Exercise #4 <br>
<p>Use the reduce function to multiply the numbers in the list below together with a lambda function.</p>

In [46]:
reduce(lambda num1, num2: num1 - num2, a_list)

-13

## Recursion <br>
<p>Recursion means that a function is calling itself, so it contanstly executes until a base case is reached. It will then push the returning values back up the chain until the function is complete. A prime example of recursion is computing factorials... such that 5! (factorial) is 5*4*3*2*1 which equals 120.</p>

#### Implementing a Base Case

In [47]:
10*9*8*7*6*5*4*3*2*1


3628800

#### Writing a Factorial Function

In [51]:
def factorial(number):
    if number <= 1:
        return number
    return number * factorial(number - 1)



In [52]:
factorial(5)

120

#### In-Class Exercise #5 <br>
<p>Write a recursive function that subtracts all numbers to the argument given.</p>

In [62]:
def subtractorial(number):
    if number <= 1:
        return number
    return number - subtractorial(number - 1)
print(subtractorial(5))

3


## Generators <br>
<p>Generators are a type of iterable, like lists or tuples. They do not allow indexing, but they can still be iterated through with for loops. They are created using functions and the yield statement.</p>

#### Yield Keyword <br>
<p>The yield keyword denotes a generator, it doesn't return so it won't leave the function and reset all variables in the function scope, instead it yields the number back to the caller.</p>

In [64]:
my_nums = list(range(1,101))

In [95]:
def sqrt_nums(a_list):
    for num in a_list:
        yield num ** .5
        
    return result

In [69]:
my_nums_sqrt = sqrt_nums(my_nums)

In [92]:
next(my_nums_sqrt)

4.795831523312719

In [96]:
for sqrt_nums in my_nums_sqrt:
    print(sqrt_nums)

1.0
1.4142135623730951
1.7320508075688772
2.0
2.23606797749979
2.449489742783178
2.6457513110645907
2.8284271247461903
3.0
3.1622776601683795
3.3166247903554
3.4641016151377544
3.605551275463989
3.7416573867739413
3.872983346207417
4.0
4.123105625617661
4.242640687119285
4.358898943540674
4.47213595499958
4.58257569495584
4.69041575982343
4.795831523312719
4.898979485566356
5.0
5.0990195135927845
5.196152422706632
5.291502622129181
5.385164807134504
5.477225575051661
5.5677643628300215
5.656854249492381
5.744562646538029
5.830951894845301
5.916079783099616
6.0
6.082762530298219
6.164414002968976
6.244997998398398
6.324555320336759
6.4031242374328485
6.48074069840786
6.557438524302
6.6332495807108
6.708203932499369
6.782329983125268
6.855654600401044
6.928203230275509
7.0
7.0710678118654755
7.14142842854285
7.211102550927978
7.280109889280518
7.3484692283495345
7.416198487095663
7.483314773547883
7.54983443527075
7.615773105863909
7.681145747868608
7.745966692414834
7.810249675906654
7.

In [93]:
my_nums_sqrt = list(sqrt_nums(my_nums))

In [94]:
my_nums_sqrt

[1.0,
 1.4142135623730951,
 1.7320508075688772,
 2.0,
 2.23606797749979,
 2.449489742783178,
 2.6457513110645907,
 2.8284271247461903,
 3.0,
 3.1622776601683795,
 3.3166247903554,
 3.4641016151377544,
 3.605551275463989,
 3.7416573867739413,
 3.872983346207417,
 4.0,
 4.123105625617661,
 4.242640687119285,
 4.358898943540674,
 4.47213595499958,
 4.58257569495584,
 4.69041575982343,
 4.795831523312719,
 4.898979485566356,
 5.0,
 5.0990195135927845,
 5.196152422706632,
 5.291502622129181,
 5.385164807134504,
 5.477225575051661,
 5.5677643628300215,
 5.656854249492381,
 5.744562646538029,
 5.830951894845301,
 5.916079783099616,
 6.0,
 6.082762530298219,
 6.164414002968976,
 6.244997998398398,
 6.324555320336759,
 6.4031242374328485,
 6.48074069840786,
 6.557438524302,
 6.6332495807108,
 6.708203932499369,
 6.782329983125268,
 6.855654600401044,
 6.928203230275509,
 7.0,
 7.0710678118654755,
 7.14142842854285,
 7.211102550927978,
 7.280109889280518,
 7.3484692283495345,
 7.416198487095663,

#### Infinite Generator

In [ ]:
# bad, never create infinite loops

#### In-Class Exercise #6 <br>
<p>Create a generator that takes a number argument and yields that number squared, then prints each number squared until zero is reached.</p>

In [122]:
my_nums = list(range(1,101))

In [123]:
def sqr_nums(a_list):
    for num in a_list:
        yield num ** 2
        
    return result

In [124]:
my_nums_sqr = sqr_nums(my_nums)

In [193]:
next(my_nums_sqr)

4761

# Exercises

### Exercise #1 <br>
<p>Filter out all of the empty strings from the list below</p>

`Output: ['Argentina', 'San Diego', 'Boston', 'New York']`

In [199]:
places = [" ","Argentina", " ", "San Diego","","  ","","Boston","New York"]

no_spaces = list(filter(lambda city: city >= ',' ,places))
print(no_spaces)

['Argentina', 'San Diego', 'Boston', 'New York']


### Exercise #2 <br>
<p>Write an anonymous function that sorts this list by the last name...<br><b>Hint: Use the ".sort()" method and access the key"</b></p>

`Output: ['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']`

In [5]:
author = ["Joel Carter", "Victor aNisimov", "Andrew P. Garfield","David hassELHOFF","Gary A.J. Bernstein"]

x = sorted(author, key= lambda name: name.split(" ")[-1].title())

print(x)

['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']


SyntaxError: unexpected EOF while parsing (1684459244.py, line 3)

### Exercise #3 <br>
<p>Convert the list below from Celsius to Farhenheit, using the map function with a lambda...</p>

`Output: [('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]
`

In [226]:
places = [('Nashua',32),("Boston",12),("Los Angelos",44),("Miami",29)]

new_places, temps = map(list, zip(*places))

print(new_places)  
print(temps)

['Nashua', 'Boston', 'Los Angelos', 'Miami']
[32, 12, 44, 29]


In [227]:
# F = (9/5)*C + 32

celcius_lamb = list(map(lambda place,temp : (place, float(temp) * (9/5) + float(32)),new_places, temps))
print(celcius_lamb)

[('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]


### Exercise #4 <br>
<p>Write a recursion function to perform the fibonacci sequence up to the number passed in.</p>

`Output for fib(5) => 
Iteration 0: 1
Iteration 1: 1
Iteration 2: 2
Iteration 3: 3
Iteration 4: 5
Iteration 5: 8`

In [4]:


def fibonacci(num):
    if num <= 1:
        return 1
    else:
        return(fibonacci(num-1) + fibonacci(num-2))
    
print(fibonacci(5))
print(fibonacci(4))
print(fibonacci(3))
print(fibonacci(2))
print(fibonacci(1))
print(fibonacci(0))

8
5
3
2
1
1
